src/test/java/io/aiven/connect/elasticsearch/ElasticsearchSinkFailureTest.java

```diff
@@ -0,0 +1,192 @@
+import io.aiven.connect.elasticsearch.bulk.BulkResponse;
+import io.aiven.connect.elasticsearch.jest.JestElasticsearchClient;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import io.searchbox.action.Action;
+import io.searchbox.client.JestClient;
+import io.searchbox.client.JestResult;
+import io.searchbox.client.JestResultHandler;
+import io.searchbox.core.BulkResult;
+import org.junit.Test;
+
+import static org.junit.Assert.fail;
+
+public class ElasticsearchSinkFailureTest {
+
+    private static class MockedJestClient implements JestClient {
+        private final BulkResult success;
+
+        public MockedJestClient(final BulkResult success) {
+            this.success = success;
+        }
+
+        @Override
+        public void close() throws IOException {
+        }
+
+        @Override
+        public <T extends JestResult> T execute(final Action<T> clientRequest) throws IOException {
+            return (T) success;
+        }
+
+        @Override
+        public <T extends JestResult> void executeAsync(final Action<T> clientRequest,
+                                                        final JestResultHandler<? super T> jestResultHandler) {
+            jestResultHandler.completed((T) success);
+        }
+
+        @Override
+        public void shutdownClient() {
+
+        }
+
+        @Override
+        public void setServers(final Set<String> servers) {
+
+        }
+    }
+
+    abstract class CustomJestElasticsearchClient extends JestElasticsearchClient {
+        public final AtomicInteger apiCallCounter = new AtomicInteger(0);
+
+        public CustomJestElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public Version getVersion() {
+            return Version.ES_V7;
+        }
+
+        @Override
+        public JsonObject getMapping(final String index, final String type) {
+            return new JsonObject();
+        }
+    }
+
+    class CustomJestTemporarilyFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestTemporarilyFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            final int numAttempt = apiCallCounter.incrementAndGet();
+
+            if (numAttempt < 3) {
+                return BulkResponse.failure(true, "");
+            }
+
+            return BulkResponse.success();
+        }
+    }
+
+    class CustomJestFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            apiCallCounter.incrementAndGet();
+            return BulkResponse.failure(false, "");
+        }
+    }
```

Comment:
Feels like we might be able to reduce some of the boilerplate introduced by these custom classes by using Mockito instead; did you consider that option?